include/fiction/algorithms/simulation/sidb/occupation_probability_excited_states.hpp

```diff
@@ -0,0 +1,118 @@
+//
+// Created by Jan Drewniok on 06.02.23.
+//
+
+#ifndef FICTION_OCCUPATION_PROBABILITY_EXCITED_STATES_HPP
+#define FICTION_OCCUPATION_PROBABILITY_EXCITED_STATES_HPP
+
+#include "fiction/algorithms/simulation/sidb/calculating_energy_and_state_type.hpp"
+#include "fiction/utils/math_utils.hpp"
+
+#include <algorithm>
+#include <cassert>
+#include <cmath>
+#include <limits>
+#include <map>
+#include <numeric>
+#include <utility>
+#include <vector>
+
+namespace fiction
+{
+
+/**
+ * This function computes the occupation probability of erroneous charge distributions (output charge does not match the
+ * expected output according the truth table) at a given temperature.
+ *
+ * @param energy_distribution This contains the energies of all possible charge distributions with the
+ * information if corresponding state is transparent or erroneous.
+ * @param temperature System temperature to assume.
+ * @return The occupation probability of all erroneous states is returned.
+ */
+[[nodiscard]] inline double occupation_probability_gate_based(const energy_and_state_type& energy_distribution,
+                                                              const double                 temperature) noexcept
+{
+    assert(!energy_distribution.empty() && "vector is empty");
+    assert((temperature > static_cast<double>(0)) && "temperature should be slightly above 0 K");
+
+    auto min_energy = std::numeric_limits<double>::infinity();
+
+    // Determine the minimal energy.
+    for (const auto& [energy, degeneracy] : energy_distribution)
+    {
+        if (energy < min_energy)
+        {
+            min_energy = energy;
+        }
+    }
```

Comment:
This could be "modernized" with `std::min_element`

Reply:
Even if we an unordered map?

Reply:
Yes. Each of your elements would then be a pair that you can unpack just like you did here with the structured binding.

Reply:
ah, I see. Thanks